Fix some whitespace problems in the docstrings

moble opened the pull request:
There are some places in the docstrings where code blocks (starting with `>>>`) follow descriptions immediately.  This seems to break the functionality of the SymPy Live blocks.  Moreover, the resulting code blocks are placed inside html `<p>` elements, meaning that the newlines are not preserved, and the formatting is just generally awful.

I used [this](https://gist.github.com/moble/58fc6b1956caa6c944bb) little script to insert the relevant newlines.  I believe this has kept the new whitespace to a minimum.  Looking at the diffs, it doesn't seem to have triggered any false positives.  In particular, it doesn't split code blocks up into separate code blocks when there is output from some of the lines.

I also took the opportunity to fix up some docstrings that were otherwise ugly in `liealgebras`, and one in `exprtools.py`.

Closes #9132.

moble committed:
Basic substitutions

moble committed:
More sensitive detection of problems

moble committed:
Clean up some of the docstrings for Lie algebras

moble committed:
Remove stray whitespace

toolforger commented on sympy/core/exprtools.py:
```diff
@@ -1299,7 +1299,9 @@ def factor_nc(expr):
     """Return the factored form of ``expr`` while handling non-commutative
     expressions.
 
-    **examples**
```
Good catch.
I did some spot testing, and it seems that the standard way to annotate "Examples" is with ```========```, so it would be nice to have that in line 1303.

moble replied:
Yeah, or `--------`.  And neither one needs an extra newline before I code block, so I didn't do that automatically.  (This one and the `liealgebra` scripts were my only manual changes.)  So there were lots of places I didn't add a newline.  But this does complicate the regex a little, and may make it more susceptible to false negatives.  So if this is converted to a code-quality check, it may be better to just add those newlines.

toolforger replied:
I meant to say "oh while you're at fixing stuff manually, === is more common than --- in this case".
I fully agree that making that kind of correction automatic is probably overeager.

moble replied:
Ah, I hadn't realized.  :)  I was also mistaken about the relative frequency of `---` versus `===`.  Apparently, the former is far more common among headings immediately before code blocks, but the latter is *way* more common among headings separated by an extra newline -- which are *way* more common in general.

skirpichev replied:
> I meant to say "oh while you're at fixing stuff manually, === is more common than --- in this case".

@toolforger, did you realize that `===` and `---` - headings of different level?

toolforger commented:
I'm impressed about the extensive docs you added to the script that you used to do this ( https://gist.github.com/moble/58fc6b1956caa6c944bb ).
If we decide to do so, these docs should make it easy to convert it to a quality check.

moble committed:
Change to more common heading delimiter

moble commented:
Moving this topic out of that line-comment thread:

I'm having doubts about my decision not to add extra newlines after `Examples` headings.  There are roughly 76 occurrences of something like this in the code:

```
Examples
========
>>> x = y**2
```

On the other hand, there are about 2496 occurrences of something like this:

```
Examples
========

>>> x = y**2
```

Note the extra newline.  So the latter are ~33 times more common.  On those grounds alone, it might be a good idea to just add the newline.  @toolforger says he thinks "making that kind of correction automatic is probably overeager", so I won't do it for now.

But my real worry is that something like this might creep into the code:

```
Examples
========
Here's an example
>>> x = y**2
```

But this would evidently lead to the original problem.  Building a regex to detect this problem while allowing the top example would be delicate.  I've checked that this doesn't happen anywhere currently, but it's certainly reasonable to expect that it could.

So if this is turned into a quality check, I would suggest just demanding that extra newline.  It wouldn't be too much of a change anyway, and would be far less bug-prone.  And if that is desired, I could do it with a trivial change to my script.

moble commented:
Aw, never mind.  I've updated [my gist](https://gist.github.com/moble/58fc6b1956caa6c944bb) with a regex that could take care of it.

toolforger commented:
Ah, I think I misunderstood.
If the regex becomes simpler *and* is more in line with what's usual in SymPy, by any means follow that route.

What I don't know is whether this should become a quality check in the long term.

moble committed:
Add some unnecessary newlines for consistency

moble commented:
Okay.  I'm happier with that.  The regex is definitely less bug prone.  Plus it just generally sets a better example to encourage people to add those newlines before code blocks in their docstrings.

moble commented:
And now Travis is failing because of bad connections to github, apparently...

moble committed:
Fix a few missed code blocks in short docstrings

It's also possible for the docstring to be completely run together, so that the
text block begins the docstring without two consecutive newlines, and is still
not separated from the code block by two newlines.  I've updated my gist to
find such instances, and applied that here.

smichr commented on bin/generate_test_list.py:
```diff
@@ -38,6 +38,7 @@ def get_paths(level=15):
     Generates a set of paths for testfiles searching.
 
     Example:
```
To be properly typeset with sphinx I believe this has to be

```
Examples  (note plural)
========
          (note blank)
>>> ...
```

moble replied:
Okay.  I hadn't been changing those because I wanted to keep a light touch, but that won't be hard.

moble committed:
Reformat certain "Example" headings with bad formatting

Following smichr's advice, I'm replacing the various improperly formatted
"Example" headings to be plural, have no colon, and be underscored by
`=========`.

moble commented:
How's that look @smichr?  Is there anything else I need to do on this?

smichr commented:
Looks a lot better to me. I am +1 for the PR.

toolforger commented:
+1 too - I didn't check every single change, but what I saw was an overall improvement.

moble commented:
Thanks.  So is there someone I should ping to get it merged, or is the typical merge timescale a bit longer?  (Sorry if I'm being dumb; just trying to get a handle on the etiquette.  :)

debugger22 commented:
I think most of my comments here https://github.com/sympy/sympy/pull/9215 have also been addressed here.

debugger22 commented:
I'll merge this in some time if there's no objection.

moble commented:
Thanks, @debugger22!
